dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/period/DefaultPeriodService.java

```diff
@@ -284,6 +296,53 @@ public Period reloadPeriod( Period period )
     {
         return periodStore.reloadForceAddPeriod( period );
     }
+
+    /**
+     * Fix issue DHIS2-7539
+     * If period doesn't exist in cache and database.
+     * Need to add and sync with database right away in a separate session/transaction.
+     * Otherwise will get foreign key constraint error in subsequence calls of batch.flush()
+     **/
+    @Override
+    @Transactional(readOnly = true)
+    public Period reloadIsoPeriodInStatelessSession( String isoPeriod )
+    {
+        Period period = PeriodType.getPeriodFromIsoString( isoPeriod );
+
+        if ( period == null )
+        {
+            return null;
+        }
+
+        Period reloadedPeriod = periodStore.reloadPeriod( period );
+
+        if ( reloadedPeriod != null )
+        {
+            return reloadedPeriod;
+        }
+
+        StatelessSession session = sessionFactory.openStatelessSession();
+        session.beginTransaction();
+        try
+        {
+            period.setPeriodType( reloadPeriodType( period.getPeriodType() ) );
+
+            session.insert( period );
+
+            return period;
+        }
+        catch ( Exception exception )
+        {
+            log.error( DebugUtils.getStackTrace( exception ) );
+        }
+        finally
+        {
+            session.getTransaction().commit();
+            session.close();
```

Comment:
Ideally, the close statement should have its own try-finally block, as the commit operation might also fail, leaving the session open.

Reply:
Agree, fixed by adding method `DbmsUtils.closeStatelessSession( session );`